standalone/phost/src/msg_sync.rs

```diff
@@ -161,46 +164,53 @@ impl<'a> MsgSync<'a> {
         Ok(())
     }
 
-    pub async fn maybe_sync_lottery_egress(&mut self, sequence: &mut u64) -> Result<()> {
-        let query_resp = self.pr.query(7, ReqData::PendingLotteryEgress {sequence: *sequence}).await?;
-        let msg_data = match query_resp {
-            QueryRespData::PendingLotteryEgress { lottery_queue_b64 } =>
-                base64::decode(&lottery_queue_b64)
-                    .map_err(|_| Error::FailedToDecode)?,
-            _ => return Err(anyhow!(Error::FailedToDecode))
-        };
+    pub async fn maybe_sync_mq_egress(&mut self) -> Result<()> {
+        // Send the query
+        let resp = self
+            .pr
+            .req_decode("get_egress_messages", GetEgressMessagesReq {})
+            .await?;
+        let messages_scl = base64::decode(&resp.messages).map_err(|_| Error::FailedToDecode)?;
+        let messages: Vec<(MessageOrigin, Vec<SignedMessage>)> =
+            Decode::decode(&mut &messages_scl[..]).map_err(|_| Error::FailedToDecode)?;
 
-        let queue: Vec<phala_types::messaging::SignedMessage> = Decode::decode(&mut &msg_data[..])
-            .map_err(|_|Error::FailedToDecode)?;
         // No pending message. We are done.
-        if queue.is_empty() {
+        if messages.is_empty() {
             return Ok(());
         }
 
         self.maybe_update_signer_nonce().await?;
 
-        let mut next_seq = *sequence;
-        for msg in &queue {
-            let msg_seq = msg.sequence;
-            if msg_seq < *sequence {
-                println!("Lottery {} has been submitted. Skipping...", msg_seq);
-                continue;
-            }
-            next_seq = cmp::max(next_seq, msg_seq + 1);
-            let ret = self.client.submit(runtimes::phala::SyncLotteryMessageCall {
-                _runtime: PhantomData,
-                msg: msg.encode()
-            }, self.signer).await;
-
-            if let Err(err) = ret {
-                println!("Failed to submit tx: {:?}", err);
-                // TODO: Should we fail early?
+        for (sender, messages) in messages {
+            let min_seq = fetch_mq_ingress_seq(self.client, sender.clone()).await?;
```

Comment:
This could be problematic. The submitted transaction it supposed to change `mq_ingress_seq`. But in reality, the tx can have delay, and can be rolled back in some rare case.

Let's be careful. The old implementation also has similar issue. We don't need to solve this problem in this PR, but we should design a correct mechanism to handle it anyway.

Life of a tx:

1. When you submit a tx, it goes to a "memory pool" first (when `.submit()` returns). Only the most basic error like empty account can be detected.
2. Then the tx from the memory pool will be selected and packed into a block, sequentialize the transactions.
    - After the block is made, node will execute the tx, and check if it succeed or not
    - Be careful: multiple nodes can propose different blocks simultaneously, resulting some "forks"
3. A consensus algorithm runs on the new block candidates, and select one to be "finalize" (usually in 2 blocks)
    - Only one block can win the election, and finalized blocks will never be rolled-back
    - In other words, some unfinalized block candidates can also lose the election

Suppose we submit a message, it gets immediately included in the mempool. Although there's no guarantee the tx will be finalized, but it's always good to assume so, and we should avoid submitting the same message to the node.

.. except some edge cases:

- The mempool may drop the tx when there's block congestion.
- The tx is included in an unfinalized block first. And at the same time, we can see "mq_ingress_seq" updated. But then a competing block wins the finalization election, resulting the unfinalized block "gets reverted" (in the view of the local node, also called "chain reorganization"). In this case, the client can even see "mq_ingress_seq" decreasing.

Usually it's safe to assume nodes will try their best to get the mempool txs finalized. In the case of a reorg, the node should pick up the missing tx, and try to include it in the next block, unless there's a serious block congestion.

Btw, each transaction has a expiration time. When a congestion happens, the block can only live in the mempool for some time (usually 3-15 mins). A too old tx will be dropped.

Reply:
Left an issue #232.